web/src/engine/osk/src/input/gestures/browser/tabletPreview.ts

```diff
@@ -0,0 +1,114 @@
+import { KeyElement } from '../../../keyElement.js';
+import KeyTipInterface from '../../../keytip.interface.js';
+import VisualKeyboard from '../../../visualKeyboard.js';
+import { GesturePreviewHost } from '../../../keyboard-layout/gesturePreviewHost.js';
+
+const BASE_CLASS = 'kmw-keypreview';
+const OVERLAY_CLASS = 'kmw-preview-overlay';
+const BASE_ID = 'kmw-keytip';
+
+export class TabletKeyTip implements KeyTipInterface {
+  public readonly element: HTMLDivElement;
+  public key: KeyElement;
+  public state: boolean = false;
+
+  private previewHost: GesturePreviewHost;
+  private preview: HTMLDivElement;
+  private readonly vkbd: VisualKeyboard;
+
+  /**
+   * Pre-builds a reusable element to serve as a gesture-preview host for selected keys
+   * on tablet-form factor devices.  This element hovers over the keyboard, staying in
+   * place (and on top) even when the layer changes.
+   */
+  constructor(vkbd: VisualKeyboard) {
+    this.vkbd = vkbd;
+    const base = this.element=document.createElement('div');
+    base.className=BASE_CLASS;
+    base.id = 'kmw-keytip';
+
+    // The following style is critical, so do not rely on external CSS
+    base.style.pointerEvents='none';
+    base.style.display='none';
+
+    this.preview = document.createElement('div');
+    base.appendChild(this.preview);
+  }
+
+  show(key: KeyElement, on: boolean, previewHost: GesturePreviewHost) {
+    const vkbd = this.vkbd;
+    const keyLayer = key?.key.spec.displayLayer;
```

Comment:
You use `key?.key` here but then on line 55 you have `key.offsetParent` (no nullish coalescing). That suggests that line 55 could crash.